drivers/sensor/default_rtio_sensor.c

```diff
@@ -0,0 +1,435 @@
+ * a 3D channel.
+ *
+ * @param[in]  header The data header to parse
+ * @param[in]  cit The channel iterator where decoding is starting
+ * @param[out] index The index of the @p header's channel_info matching the @p cit
+ * @param[out] offset the offset of the channel value if the channel value is a 3D channel,
+ *             0 otherwise.
+ * @return 0 on success
+ * @return -EINVAL if the index couldn't be computed
+ */
+static int compute_channel_info_index(const struct sensor_data_generic_header *header,
+				      sensor_channel_iterator_t cit, int *index, int *offset)
+{
+	sensor_channel_iterator_t i = {0};
+
+	*index = 0;
+	do {
+		sensor_channel_iterator_t last_i = i;
+
+		i += SENSOR_CHANNEL_3_AXIS(header->channel_info[*index].channel) ? 3 : 1;
+
+		if (i > cit) {
+			*offset = cit - last_i;
+			return 0;
+		}
+		*index += 1;
+	} while (*index < header->num_channels);
+
+	return -EINVAL;
+}
+
+/**
+ * @brief Default decoder decode N samples
+ *
+ * Decode up to N samples starting at the provided @p fit and @p cit. The appropriate channel types
+ * and q31 values will be placed in @p values and @p channels respectively.
+ *
+ * @param[in]     buffer The data buffer to decode
+ * @param[in,out] fit The starting frame iterator
+ * @param[in,out] cit The starting channel iterator
+ * @param[out]    channels The decoded channel types
+ * @param[out]    values The decoded q31 values
+ * @param[in]     max_count The maximum number of values to decode
+ * @return > 0 The number of decoded values
+ * @return 0 Nothing else to decode on this @p buffer
+ * @return < 0 Error
+ */
+static int decode(const uint8_t *buffer, sensor_frame_iterator_t *fit,
+		  sensor_channel_iterator_t *cit, enum sensor_channel *channels, q31_t *values,
+		  uint8_t max_count)
+{
+	const struct sensor_data_generic_header *header =
+		(const struct sensor_data_generic_header *)buffer;
+	int num_samples = 0;
+	const q31_t *q =
+		(const q31_t *)(buffer + sizeof(struct sensor_data_generic_header) +
+				header->num_channels * sizeof(struct sensor_data_generic_channel));
+
+	if (*fit != 0) {
+		return -EINVAL;
+	}
+
+	for (int i = 0; i < header->num_channels; ++i) {
+		num_samples += SENSOR_CHANNEL_3_AXIS(header->channel_info[i].channel) ? 3 : 1;
+	}
+
+	int count = 0;
+	int channel_info_index;
+	int offset;
+
+	while (true) {
+		if (compute_channel_info_index(header, *cit, &channel_info_index, &offset) != 0) {
+			LOG_INF("Could not compute the channel index, channel iterator may be out "
+				"of bounds");
+			return -EINVAL;
+		}
+
+		if (header->channel_info[channel_info_index].channel == SENSOR_CHAN_MAX) {
+			*cit += 1;
+		} else {
+			break;
+		}
+	}
+
+	for (; *cit < num_samples && count < max_count; ++count) {
+		enum sensor_channel channel = header->channel_info[channel_info_index].channel;
+
+		channels[count] = channel;
+		if (SENSOR_CHANNEL_3_AXIS(channel)) {
+			channels[count] = channel - 3 + offset;
```

Comment:
See my comment on `sensor_reconfigure_read_iodev()`.  The application calls reconfigure with a single channel, but decode will return data for 3 channels.